File: utils/zoperestart.c
/*
 * $Id$ 
 */

#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>

#define TMOUT 60

int restart(const char *d){
    struct stat zopedirstat;
    FILE *log;
    char buf[1024];
    char zopectl[1024];
    char zopelog[1024];
    int ticks;
    int timeout;
    char *timeoutstring;
    char *myzopectl;
    char *myzopelog;

    /* Check that directory exists */
    if(stat(d, &zopedirstat)==-1){
        perror("stat");
        return 1;
    }

    /* Get timeout from environment */
    timeout = TMOUT;
    timeoutstring = getenv("ZOPE_RESTART_TIMEOUT");
    if(timeoutstring){
        timeout = atoi(timeoutstring);
    }

    /* Get alternative locations from environment */
    myzopectl = getenv("ZOPE_RESTART_SCRIPT");
    myzopelog = getenv("ZOPE_LOG_FILE");

    /* Set up required strings */
    if(myzopectl){
        snprintf(zopectl, sizeof(zopectl), "%s restart", myzopectl);
    } else {
        snprintf(zopectl, sizeof(zopectl), "%s/bin/zopectl restart", d);
    }

    if(myzopelog){
        snprintf(zopelog, sizeof(zopelog), "%s", myzopelog);
    } else {
        snprintf(zopelog, sizeof(zopelog), "%s/log/event.log", d);
    }

    /* Open log file and seek to the end */
    log = fopen(zopelog, "r");
    fseek(log, 0, SEEK_END);

    /* Restart the client */
    fprintf(stderr, "Calling %s\n", zopectl);
    system(zopectl);

    /* tail the logfile and wait for startup to complete */
    ticks = 0;
    while(ticks<timeout){
        /* Read log */
        if(fgets(buf, sizeof(buf), log)){
            /* Check the line */
            if(strstr(buf, "Zope Ready to handle requests")){
                fprintf(stderr, " success\n");
                return 0;
            }
        } else {
            sleep(1);
            ticks++;
            fprintf(stderr, ".");
        }
    }
    fprintf(stderr, " alarm\n");
    return 1;
}

int main(int argc, char **argv){
    int i;
    int x;

    if(argc<2){
        fprintf(stderr, "Usage: %s zopeclientdir [...]\n", argv[0]);
        return 3;
    }

    for(i=1; i<argc; i++){
        if((x=restart(argv[i]))!=0){
            return x;
        }
    }
    return 0;
}

/* vim: set ts=4 sw=4 expandtab: */

File: utils/zoperestart.py
import sys
import os
import subprocess
from time import sleep
from select import poll, POLLIN
from StringIO import StringIO
from threading import Thread, Event

class LogReaderThread(Thread):
    def __init__(self, log, monitor, signal):
        Thread.__init__(self)
        self.log = log
        self.signal = signal
        self.monitor = monitor

    def run(self):
        try:
            line = self.log.readline()
            while line:
                sys.stderr.write('.')
                if self.monitor.isSet() and \
                    line.strip().endswith("Zope Ready to handle requests"):
                    self.signal.set()
                    return
                line = self.log.readline()
        except ValueError:
            # File was closed, just roll over and die
            pass

def run(args, stdout=None):
    if stdout is None:
        stdout=subprocess.PIPE
    app = subprocess.Popen(args, stdin=subprocess.PIPE, stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT)
    # Close stdin, we won't use it
    app.stdin.close()

    return app

def restart(args, timeout):
    restartcmd = args + ["restart"]
    logtailcmd = args + ["logtail"]

    # Run logtail command, do this first to seek to the end of the current
    # log before restarting.
    sys.stderr.write(" flush ")
    tailapp = run(logtailcmd)
    restartedevent = Event()
    monitorevent = Event()
    reader = LogReaderThread(tailapp.stdout, monitorevent, restartedevent)
    reader.start()

    # Restart zope and wait for it to finish
    sys.stderr.write(" restart ")
    file13 = open('/dev/null', 'w')
    zopectl = run(restartcmd, stdout=file13)
    zopectl.wait()
    sys.stderr.write(" done ")

    # Now signal the Log reader thread to start looking
    monitorevent.set()

    # Wait for the thread to complete or timeout
    restartedevent.wait(timeout)

    # And stop the logtail command
    tailapp.terminate()

    if restartedevent.isSet():
        print >>sys.stderr, " success"
        return 0

    print >>sys.stderr, " alarm"
    return 1

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print >>sys.stderr, "Usage: %s zopectlscript" % sys.argv[0]
        sys.exit(3)

    timeout = int(os.environ.get('ZOPE_RESTART_TIMEOUT', '60'))
    for arg in sys.argv[1:]:
        x = restart([arg], timeout)
        if x != 0:
            sys.exit(x)
        sys.exit(1)
